feat: modify accept share method

ravneet-bitgo opened the pull request:
Ticket: PX-3296

## Description
We are modifying the accept share method for go account wallet invitations. If the `keychainOverrideRequired` key is present in the walletShare, we need to create a new keychain and pass the same to the `updateShare` method so we can override the existing key. After successful completion of the same, we need to `reshareWallet` to all the existing spenders.

TDD: [link](https://docs.google.com/document/d/1Dll1Z2OnVi2ofR__4LlqDdEI1aNUPuLXrqnITPQ0sf8/edit#bookmark=id.tr9x4ewh6fhv)

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

ravneet-bitgo commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -506,10 +508,7 @@ export class Wallets implements IWallets {
-      .post(this.baseCoin.url('/walletshare/' + params.walletShareId))
-      .send(params)
-      .result();
+    return await this.bitgo.post(this.baseCoin.url('/walletshare/' + params.walletShareId)).send(params);
```
We have removed `.result()` because we need to check the status code of this API later.

shivagupta94 commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+   * @param   {Number} numWords     Number of 32-bit words
+   * @returns {String}          base58 random password
+   */
+  generateRandomPassword(numWords = 5): string {
```
Lets ADD TODO comment here, if we can find some better place for this method

manav-c replied:
https://github.com/BitGo/BitGoJS/pull/4419/files#r1566982186

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/iWallet.ts:
```diff
@@ -609,6 +609,7 @@ export interface WalletData {
   tokens?: Record<string, any>[];
   nfts?: { [contractAddressOrToken: string]: NftBalance };
   unsupportedNfts?: { [contractAddress: string]: NftBalance };
+  users?: any[];
```
+1, please type this out. Even if it is not fully typed, add what known fields you are using.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+   */
+  async reshareOfcAccountWithSpenders(walletId: string, userPassword: string): Promise<void> {
+    const wallet = await this.bitgo.coin('ofc').wallets().get({ id: walletId });
+    const enterpriseUsersResponse = (await this.bitgo
```
Assert `wallet.wallet.enterprise` above.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+    const wallet = await this.bitgo.coin('ofc').wallets().get({ id: walletId });
+    const enterpriseUsersResponse = (await this.bitgo
+      .get(`/api/v2/enterprise/${wallet?._wallet?.enterprise}/user`)
+      .result()) as any;
```
Why is this typed as any?

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+          wallet.shareWallet(shareParams);
+        }
+      } catch (e) {
+        // TODO: gracefully handle this error
```
Please address this TODO.

shivagupta94 replied:
Do you suggest adding a ticket or providing a thorough resolution?

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+        }
+      } catch (e) {
+        // TODO: gracefully handle this error
+        console.error(e);
```
Reove this console.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+   * @returns
+   */
+  async createKeychain(userPassword: string): Promise<Keychain> {
+    const sdkCoin = await this.bitgo.coin('ofc');
+    const keychains = sdkCoin.keychains();
```
```suggestion
    const keychains = this.baseCoin.keychains();
```

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +627,28 @@ export class Wallets implements IWallets {
+      if (_.isUndefined(params.userPassword)) {
+        throw new Error('userPassword param must be provided to decrypt shared key');
+      }
+      const walletKeychain = await this.createKeychain(params.userPassword);
```
Why are we creating a new keychain? Shouldn't it just be an update?

shivagupta94 replied:
We only need to generate a new keychain when keychainOverrideRequired is set to true, indicating the intentional need to create a new keychain and replace the existing one in the wallet.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +627,28 @@ export class Wallets implements IWallets {
+      // If the wallet share was accepted successfully (changed=true), reshare the wallet with the spenders
+      if (response.changed && response.state === 'accepted') {
+        try {
+          await this.reshareOfcAccountWithSpenders(walletShare.wallet, params.userPassword);
```
Is `keychainOverrideRequired` only set for a single wallet share on the OFC wallet? Is there a case where two users have this set to true in their wallet share?

shivagupta94 replied:
> Is there a case where two users have this set to true in their wallet share?

Yes, it's possible for two different wallet share requests from different users to have `keychainOverrideRequired` set to true. However, it's important to note that only the Admin APIs, accessible exclusively to the support team, can set this `keychainOverrideRequired` flag.

shivagupta94 replied:
So ideally, it's the responsibility of the support team to ensure that only one invitation is sent at a time.

manav-c replied:
As a follow up, we can come back and put some logic in the admin API (that sends out the invitation) to abort if there is already an active WalletShare for a wallet with the `keychainOverrideRequired` property set to true.

manav-c replied:
> So ideally, it's the responsibility of the support team to ensure that only one invitation is sent at a time

Ideally, we should not rely on this.

zahin-mohammad replied:
Please address the concern here. If it's possible to have two admins with the keychain override, then wallet spenders will get invited multiple times when they do not need to. There should be restrictions in place to ensure only one admin at a time is doing this op.

shivagupta94 replied:
@rajucse fyi^

shivagupta94 replied:
@zahin-mohammad 
I had a discussion with the team [here](https://bitgo.slack.com/archives/C06MHA22P08/p1713423281043779) regarding the concern you raised. The team has decided to prioritize picking up the improvement later, and for the time being, we will proceed with the current approach.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +627,28 @@ export class Wallets implements IWallets {
+        try {
+          await this.reshareOfcAccountWithSpenders(walletShare.wallet, params.userPassword);
+        } catch (e) {
+          // TODO: gracefully handle this error
+          console.error(e);
```
Address these TODO's

shivagupta94 replied:
Do you suggest adding a ticket or providing a thorough resolution?

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,77 @@ export class Wallets implements IWallets {
+   * @param walletId
+   * @param userPassword
+   */
+  async reshareOfcAccountWithSpenders(walletId: string, userPassword: string): Promise<void> {
```
Are these methods meant to be exposed? If so they need to updated in the wallet interface in `iWallet.ts`

zahin-mohammad reviewed (changes requested):
Please add tests

manav-c commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +542,75 @@ export class Wallets implements IWallets {
+   * @param   {Number} numWords     Number of 32-bit words
+   * @returns {String}          base58 random password
+   */
+  generateRandomPassword(numWords = 5): string {
```
why are you redefining this? it already exists in the SDK - https://github.com/BitGo/BitGoJS/blob/97c0f3ce5aa83bd4d412ba278be931473bc7c923/modules/sdk-api/src/bitgoAPI.ts#L1106

ravneet-bitgo replied:
We are aware of this, moving it to a util method and import it at both the places because as of now the sdk-api is not imported in sdk-core (circular dependency issue).

manav-c commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +542,75 @@ export class Wallets implements IWallets {
+      input: newKeychain.prv,
+    });
+
+    const walletKeychain = await keychains.add({
```
is the keychain also being added as part of the `updateShare` flow? @shivagupta94

shivagupta94 replied:
> is the keychain also being added as part of the updateShare flow? @shivagupta94
You mean in the wallet?

So, we need to create a keychain on the client side and save it in the backend using `keychains.add` method. Then, when triggering the updateShare method, we should pass the newly generated keychain ID in the payload. This way, we can replace it with the existing key ID.

manav-c replied:
No. To the DB. This step - https://docs.google.com/document/d/1Dll1Z2OnVi2ofR__4LlqDdEI1aNUPuLXrqnITPQ0sf8/edit#bookmark=id.a9tdqvm0ql4f

shivagupta94 replied:
We're using the `keychains.add` function for this purpose. It saves the keychain details in the database and provides us with a reference ID for the same.

shivagupta94 replied:
> I was just asking if you're doing this during updateShare too and you're not

Yes we are adding keychain to DB on client side only

> if updateShare fails for any reason, then this key will be unused (not on any wallet)

I agree, implementing that approach would require setting up a rollback mechanism as well. Since creating a keychain is a prerequisite for this acceptshare flow, ensuring we can revert any erroneous actions would be essential.

manav-c replied:
> I agree, implementing that approach would require setting up a rollback mechanism as well

You can use mongo transactions in WP for this

manav-c commented on modules/sdk-core/src/bitgo/utils/ofc.ts:
```diff
@@ -0,0 +1,12 @@
+ * @param   {Number} numWords     Number of 32-bit words
+ * @returns {String}          base58 random password
+ */
+export function generateRandomPassword(numWords: number): string {
```
is this just an ofc util?

manav-c commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +617,27 @@ export class Wallets implements IWallets {
+        try {
+          await this.reshareOfcAccountWithSpenders(walletShare.wallet, params.userPassword);
+        } catch (e) {
+          debug('failed to reshare wallet with spenders', e);
```
we should ideally be very loud in this case and raise alarms because this means current spenders on the wallet cannot sign txns (due to the key being replaced) and manual intervention is required. Someone should be notified.

shivagupta94 replied:
I agree. Initially, we discussed this, but since we don't have any features available on the client side to capture these errors and push them to the backend or raise any kind of alarm, we have to log them only.

However, we could potentially create a new API that can consume these errors and notify us about such cases, allowing us to take necessary steps.

Alternatively, we could consider throwing the error instead of consuming it. This way, we can inform newly added users that no spender can perform any transaction due to this issue. @manav-c , what are your thoughts on this approach?

manav-c replied:
Yeah I guess we could abort from this flow so the invited user has to accept their invite again

zahin-mohammad replied:
Please throw the error in this case. Debug logs will get swalloed by the client side and should not exist in our production sdk code.

shivagupta94 replied:
I had a discussion with the team [here](https://bitgo.slack.com/archives/C06MHA22P08/p1713423281043779) regarding the concern you raised. The team has decided to prioritize picking up the improvement later, and for the time being, we will proceed with the current approach.

zahin-mohammad replied:
This is not acceptible for production code. Please throw and remove the debug logs.

manav-c commented:
please add tests

mmcshinsky-bitgo commented on modules/sdk-core/src/bitgo/wallet/iWallet.ts:
```diff
@@ -575,6 +575,11 @@ export interface CrossChainUTXO {
 export type WalletType = 'backing' | 'cold' | 'custodial' | 'custodialPaired' | 'hot' | 'trading';
 export type SubWalletType = 'distributedCustody';
 
+export type WalletUser = {
```
Can we pull this from `@bitgo/public-types`. If not, we may want to add it to provide full types.

zahin-mohammad replied:
+1

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +615,27 @@ export class Wallets implements IWallets {
-
+    const walletShare = await this.getShare({ walletShareId: params.walletShareId });
+    if (walletShare.keychainOverrideRequired && walletShare.permissions.indexOf('admin') !== -1) {
+      if (_.isUndefined(params.userPassword)) {
+        throw new Error('userPassword param must be provided to decrypt shared key');
+      }
```
Throwing here is misleading. Encrypted key data is only shared with people with `spend` permissions, not `admin`. You can technically be an admin and not a spender.

zahin-mohammad replied:
> they will have encrypted key data, which holds true for two roles: admin and spenders.

This is not true for the amdin role. There is no gaurentee that the `admin` has the prv material. 

You need to check both admin and spender permissions here, not just admin.

shivagupta94 replied:
Done

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,65 @@ export class Wallets implements IWallets {
+   * @param walletId
+   * @param userPassword
+   */
+  async reshareOfcAccountWithSpenders(walletId: string, userPassword: string): Promise<void> {
```
```suggestion
  async reshareWalletWithSpenders(walletId: string, walletPassword: string): Promise<void> {
```

ravneet-bitgo replied:
This is the userPassword and not walletPassword.

<img width="1728" alt="Screenshot 2024-04-18 at 12 36 19 PM" src="https://github.com/BitGo/BitGoJS/assets/131337791/74af269a-64fd-47b5-bd39-6586f0c766ba">

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,65 @@ export class Wallets implements IWallets {
+   * @param userPassword
+   */
+  async reshareOfcAccountWithSpenders(walletId: string, userPassword: string): Promise<void> {
+    const wallet = await this.bitgo.coin('ofc').wallets().get({ id: walletId });
```
```suggestion
    const wallet = await this.get({ id: walletId });
```

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,65 @@ export class Wallets implements IWallets {
+    const usersMap = new Map(
+      [...enterpriseUsersResponse?.adminUsers, ...enterpriseUsersResponse?.nonAdminUsers].map((obj) => [obj.id, obj])
+    );
+    wallet?._wallet?.users?.forEach(async (user) => {
+      // user should be a spender and not an admin
+      if (user.permissions.includes('spend') && !user.permissions.includes('admin')) {
+        const userObject = usersMap.get(user.user);
+        const shareParams = {
+          walletId: walletId,
+          user: user.user,
+          permissions: user.permissions.join(','),
+          walletPassphrase: userPassword,
+          email: userObject.email.email,
+          coin: wallet.coin,
+          reshare: true,
+        };
+        wallet.shareWallet(shareParams);
+      }
+    });
```
Consider using `Promise.all` in conjunction with a `map`. Map all wallet users to a promise, then await them all.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -538,6 +540,65 @@ export class Wallets implements IWallets {
+   * @param userPassword
+   * @returns
+   */
+  async createKeychain(userPassword: string): Promise<Keychain> {
```
This should live in the `keychains` class if its not already there.

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +615,27 @@ export class Wallets implements IWallets {
+      // If the wallet share was accepted successfully (changed=true), reshare the wallet with the spenders
+      if (response.changed && response.state === 'accepted') {
+        try {
+          await this.reshareOfcAccountWithSpenders(walletShare.wallet, params.userPassword);
```
```suggestion
          await this.reshareWalletWithSpenders(walletShare.wallet, params.userPassword);
```

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -28,6 +28,8 @@ import {
 } from './iWallets';
 import { Wallet } from './wallet';
 
+const debug = require('debug')('bitgo:v2:wallets');
```
```suggestion
```

zahin-mohammad replied:
Please remove this.

shivagupta94 replied:
Done

zahin-mohammad reviewed (changes requested):
Have some suggestions/concerns on the implementation here. Please see the review comments.

shivagupta94 commented:
> Please add tests

Added

zahin-mohammad commented on modules/sdk-core/src/bitgo/keychain/keychains.ts:
```diff
@@ -427,4 +427,28 @@ export class Keychains implements IKeychains {
+   * @param userPassword
+   * @returns
+   */
+  async createWalletKeychain(userPassword: string): Promise<Keychain> {
```
```suggestion
  async createUserKeychain(userPassword: string): Promise<Keychain> {
```

zahin-mohammad replied:
Its not technically a wallet keychain until you associate the key with the wallet. Keychains are standalone objects.

shivagupta94 replied:
Done

zahin-mohammad commented on modules/sdk-core/src/bitgo/keychain/keychains.ts:
```diff
@@ -427,4 +427,28 @@ export class Keychains implements IKeychains {
+      input: newKeychain.prv,
+    });
+
+    const walletKeychain = await keychains.add({
+      encryptedPrv,
+      originalPasscodeEncryptionCode,
+      pub: newKeychain.pub,
+      source: 'user',
+    });
+    return walletKeychain;
```
```suggestion
    return await keychains.add({
      encryptedPrv,
      originalPasscodeEncryptionCode,
      pub: newKeychain.pub,
      source: 'user',
    });
```

shivagupta94 replied:
Done

zahin-mohammad commented on modules/sdk-core/src/bitgo/keychain/iKeychains.ts:
```diff
@@ -181,4 +181,5 @@ export interface IKeychains {
   getKeysForSigning(params?: GetKeysForSigningOptions): Promise<Keychain[]>;
   createMpc(params: CreateMpcOptions): Promise<KeychainsTriplet>;
   createTssBitGoKeyFromOvcShares(ovcOutput: OvcToBitGoJSON): Promise<BitGoKeyFromOvcShares>;
+  createWalletKeychain(userPassword: string): Promise<Keychain>;
```
```suggestion
  createUserKeychain(userPassword: string): Promise<Keychain>;
```

shivagupta94 replied:
Done

zahin-mohammad commented on modules/sdk-core/src/bitgo/wallet/wallets.ts:
```diff
@@ -554,9 +594,27 @@ export class Wallets implements IWallets {
-    const walletShare = (await this.getShare({ walletShareId: params.walletShareId })) as any;
-
+    const walletShare = await this.getShare({ walletShareId: params.walletShareId });
+    if (walletShare.keychainOverrideRequired && walletShare.permissions.indexOf('admin') !== -1) {
```
```suggestion
    if (walletShare.keychainOverrideRequired && walletShare.permissions.indexOf('admin') !== -1 && walletShare.permissions.indexOf('spend') !== -1) {
```

shivagupta94 replied:
Done

shivagupta94 committed:
feat(sdk-core): modify accept share method

Ticket: PX-3296
We are modifying the accept share method for go account wallet invitations
